File: hackerRank/data_structures/InsertANodeAtTheHeadOfALinkedList/insert-a-node-at-the-head-of-a-linked-list.py
# https://www.hackerrank.com/challenges/insert-a-node-at-the-head-of-a-linked-list/problem
class SinglyLinkedListNode:
    def __init__(self, node_data):
        self.data = node_data
        self.next = None


class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None


def print_singly_linked_list(node, sep):
    while node:
        print(node.data, sep=sep)
        node = node.next


def insertNodeAtHead(llist, data):
    new_ll = SinglyLinkedListNode(data)
    new_ll.next = llist
    return new_ll


if __name__ == '__main__':
    n = int(input())

    ll = SinglyLinkedList()

    for _ in range(n):
        item = int(input())
        ll.head = insertNodeAtHead(ll.head, item)

    print_singly_linked_list(ll.head, '\n')

File: hackerRank/data_structures/InsertANodeAtTheHeadOfALinkedList/InsertANodeAtTheHeadOfALinkedList.java
package data_structures.InsertANodeAtTheHeadOfALinkedList;

import java.io.IOException;
import java.util.Scanner;

// https://www.hackerrank.com/challenges/insert-a-node-at-the-head-of-a-linked-list/problem
public class InsertANodeAtTheHeadOfALinkedList {
    static class SinglyLinkedListNode {
        public int data;
        public SinglyLinkedListNode next;

        public SinglyLinkedListNode(int nodeData) {
            this.data = nodeData;
            this.next = null;
        }
    }

    static class SinglyLinkedList {
        public SinglyLinkedListNode head;
        public SinglyLinkedListNode tail;

        public SinglyLinkedList() {
            this.head = null;
            this.tail = null;
        }
    }

    public static void printSinglyLinkedList(SinglyLinkedListNode node, String sep) throws IOException {
        while (node != null) {
            System.out.print(node.data);
            node = node.next;

            if (node != null) {
                System.out.print(sep);
            }
        }
    }

    static SinglyLinkedListNode insertNodeAtHead(SinglyLinkedListNode llist, int data) {
        SinglyLinkedListNode ll = new SinglyLinkedListNode(data);
        ll.next = llist;
        return ll;
    }

    public static void main(String[] args) throws IOException {
        Scanner scanner = new Scanner(System.in);
        SinglyLinkedList ll = new SinglyLinkedList();
        int n = scanner.nextInt();

        for (int i = 0; i < n; i++) {
            int item = scanner.nextInt();
            ll.head = insertNodeAtHead(ll.head, item);
        }

        printSinglyLinkedList(ll.head, "\n");
        scanner.close();
    }
}